### DISCUSSION TITLE
Fade out + skip on the same source, plus voice + bg music

### DISCUSSION THREAD

**Discussion by ArvyRogerio** [savonet/liquidsoap#2582] [User Support]
Hey there! I'm working on a .liq script for some months, and now it's almost done (and amazing). Except for one little thing, that I still cannot do, even I've tried a lot of things :(

My script works this way, basically:
- a song is being played
- remaining 10s it reads the next song from a PHP script (to queue)
- cue_cut + liq_cross_duration metadata to overlap
- loop

Works great, but, sometimes, I want to insert a pre-recorded voiceover ("acapella"), that has not a fixed length. And I have a 5 minute bg music, that I need to fade out when the voice is over.

My ideia:

- song 1 is being played
- remaining 10s it reads the next song from a PHP script (to queue)
- the new song is detected as a bg music (metadata genre) and starts to play normally
- after 3s, using on_track I insert the voice.mp3
   _thread.run(delay=3.,fun() -> source2.push.uri(!voicefile))_
   _(before I've defined: source1 = add(normalize=false,[source2,source1]))_
- when voice.mp3 is over, it want to fade out the bg music
   fade out -> source1 skip (after ~5s of fade out)
- cue_cut + liq_cross_duration metadata to overlap
- loop to next normal song

Everything is working! But I don't know how to fade out in the *current* source, and keep on it. 

Using source.skip(source1) works perfectly, but I want to fade out the bg music before the skip...

Something like:

_def trigger_source2_fadeout() =
  get_new_song()
  *----fade---- here is the problem* 
  source.skip(source1)
  ()
end_

_def trigger_source2(m) =
  thread.run(delay=source2.remaining()+5.,trigger_source2_fadeout)
end_

and: _source2.on_track(trigger_source2)_

Any ideas?
Thanks a lot!

* Extra note: I had an idea, to set source1 volume down progressively in a 5 seconds loop (from 0db to -inf db), skip and set volume again to 0 db, but I didn't try yet. Looking for a more intelligent idea.

![image](https://user-images.githubusercontent.com/11563884/186977480-92682b4a-dcd9-41ee-80fa-13a5516b7260.png)

**Comment by toots** [comment]
Hi! Have you tried using `smooth_add`:
```ruby
s = ... # main source
voice = ... # added voice
s = smooth_add(normal=s, special=voice)
```
If the operator doesn't exactly fit your need you can look at its code and see if you can adapt it:
```ruby
# Mixes two streams, with faded transitions between the state when only the
# normal stream is available and when the special stream gets added on top of
# it.
# @category Source / Track Processing
# @flag extra
# @param ~duration Duration of the fade in seconds.
# @param ~p       Portion of amplitude of the normal source in the mix.
# @param ~normal  The normal source, which could be called the carrier too.
# @param ~special The special source.
def smooth_add(~duration=1., ~p=getter(0.2), ~normal, ~special)
  p = getter.function(p)
  last_p = ref(p())

  def c(fn,s) =
    def v() =
      fn = !fn
      fn()
    end
    fade.scale(v,s)
  end

  special_volume = ref(fun () -> 0.)
  special = c(special_volume,special)

  normal_volume = ref(fun () -> 1.)
  normal = c(normal_volume,normal)

  def to_special(_,special) =
    last_p := p()
    q = 1. - !last_p
    normal_volume := mkfade(start=1.,stop=!last_p,duration=duration,normal)
    special_volume := mkfade(stop=q,duration=duration,special)
    special
  end

  def to_blank(special,b)
    normal_volume := mkfade(start=!last_p,duration=duration,normal)
    special_volume := mkfade(start=1.-!last_p,duration=duration,special)
    sequence([special,b])
  end

  special = fallback(track_sensitive=false,
               transitions=[to_special,to_blank],
               [special,blank()])

  add(normalize=false,[normal,special])
end
```

**Reply by ArvyRogerio** [reply]
Hi @toots,

I'll try, but I think my main problem is how to "trigger" it when I manually detect (event) that the "voice.mp3" is finished...

I have some difficult to understand how to trigger something "once" and, after did, keep the previous source config.
